File: java/src/main/creational/factory/Factory.java
package creational.factory;

// Factory class that selects the right creator dynamically using enums
public class Factory {
    public static Product createProduct(ProductType type) {
        if (type == null) {  // Fix: Handle null properly
            throw new IllegalArgumentException("Product type cannot be null");
        }

        Creator creator;
        switch (type) {
            case A:
                creator = new ConcreteCreatorA();
                break;
            case B:
                creator = new ConcreteCreatorB();
                break;
            default:
                throw new IllegalArgumentException("Invalid product type: " + type);
        }
        return creator.createProduct();
    }
}


// Example Usage:
// public class FactoryMethodExample {
//     public static void main(String[] args) {
//         Product productA = Factory.createProduct(ProductType.A);
//         productA.use();
//         Product productB = Factory.createProduct(ProductType.B);
//         productB.use();
//     }
// }
//
// Common mistakes to avoid:
// 1. Tight Coupling Between Factory and Products
//    Mistake: Directly tying the factory to specific concrete classes.
//    Fix: Use abstractions (e.g., interfaces or base classes) to reduce coupling.
//
// 2. Missing Validation
//    Mistake: Not handling invalid inputs properly in the factory.
//    Fix: Always validate input and throw meaningful errors for unsupported cases.
//
// 3. Overcomplicating the Factory
//    Mistake: Adding too much logic to the factory, making it hard to maintain.
//    Fix: Keep the factory focused on object creation, and avoid business logic.
//
// 4. Hardcoding the Factory's Behavior
//    Mistake: Embedding product types as strings directly into the factory.
//    Fix: Use an enum (`ProductType`) to improve type safety and maintainability.

File: python/src/creational/factory/factory.py
from .product import Product, ConcreteProductA, ConcreteProductB
from .product_type import ProductType


class Factory:
    """
    Factory: Provides a method to create objects without
    specifying the exact class.
    """

    @staticmethod
    def create_product(product_type: ProductType) -> Product:
        """
        Factory Method: Returns an instance of a
        product based on the provided type.
        """
        if product_type == ProductType.PRODUCT_A:
            return ConcreteProductA()
        elif product_type == ProductType.PRODUCT_B:
            return ConcreteProductB()
        else:
            raise ValueError(f"Unknown product type: {product_type}")

# Example Usage:
#
# from src.creational.factory.factory import Factory
# from src.creational.factory.product_type import ProductType
#
# product_a = Factory.create_product(ProductType.PRODUCT_A)
# print(product_a.operation())  # Output: "ConcreteProductA operation executed"
#
# product_b = Factory.create_product(ProductType.PRODUCT_B)
# print(product_b.operation())  # Output: "ConcreteProductB operation executed"
#
# Common Mistakes to Avoid:
#
# 1. Hardcoding Class Names in Factory:
#    - Mistake: Using raw strings to determine product types.
#    - Fix: Use enums (`ProductType`) to avoid typos and
#    improve maintainability.
#
# 2. Allowing Factory to Grow Uncontrollably:
#    - Mistake: Adding too many `if-elif` conditions inside the factory.
#    - Fix: Use a registry pattern if new products need
#    to be added dynamically.
#
# 3. Misusing the Factory for Simple Object Creation:
#    - Mistake: Using the factory when simple instantiation
#    (`ConcreteProductA()`) is sufficient.
#    - Fix: Use Factory only when object creation logic is complex or varies.
